Merge of initial 7-BM customizations

aps-7bm opened the pull request:
I have done some initial testing of my customizations for 7-BM.  One of the main differences from the 2-BM version is that I am calculating the PSO parameters in the tomoscan_7bm.py file, rather than using the PSOFly infrastructure set up by Tim Mooney from BCDA.  The logic is based on Tim's PSOFly, and mostly sends commands directly to the Aerotech Ensemble driver through the asyn record set up for PSOFly.

aps-7bm committed:
Merge pull request #1 from tomography/master

Update to reflect changes to tomography/tomoscan

aps-7bm committed:
Adding in files from the iocBoot directory for 7-BM.

aps-7bm committed:
Merge remote-tracking branch 'upstream/master'

aps-7bm committed:
Minor changes to tomoscan_7bm

aps-7bm committed:
Merge branch 'master' of https://github.com/tomography/tomoscan

Conflicts:
	configure/RELEASE

aps-7bm committed:
Work in progress on manual PSO calculations for 7-BM.

aps-7bm committed:
Merge remote-tracking branch 'upstream/master'

aps-7bm committed:
Somewhat working version before rebuilding IOC.

aps-7bm committed:
Edit typos on template and settings.req files for 7BM

aps-7bm committed:
Some typo fixes on template and req files before nuking IOC.

aps-7bm committed:
First commit after getting things to build correctly.

Have PSO calculations in python, with PVs to control
talking to the Aerotech driver.

aps-7bm committed:
IOC boots and seems to work.

Significant edits to MEDM screen to incorporate Python PSO calcs.
Added a $(P)$(R)Testing PV to disable shutter opening for testing.

aps-7bm committed:
Recent template and Python file changes during initial testing.

aps-7bm committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	tomoscan/tomoscan.py

aps-7bm committed:
Bug fix to PSO calculations.

decarlof commented:
@aps-7bm I would remove the print() in the tomoscan.py and perhaps undo the changes in configure/RELEASE as this is happening when following the [Install directions](https://tomoscan.readthedocs.io/en/latest/install.html).
I like your PSO handling better, we should do the same at 2-BM and 32-ID and perhaps consider merging at one point 2-BM and 7-BM. 

@MarkRivers @aps-7bm please hold merging as I am adding a doc update to include the 7-BM additions (will be done later today)

decarlof committed:
added docs for 7-BM

aps-7bm committed:
Merge pull request #2 from decarlof/master

added docs for 7-BM

aps-7bm committed:
Remove print() statements from tomoscan.py used for debugging
for 7-BM.

Remove configure/RELEASE from the 7-BM repo so it will merge
appropriately with the upstream repo.

aps-7bm committed:
Merge branch 'master' of https://github.com/aps-7bm/tomoscan

aps-7bm committed:
Add the original version of configure/RELEASE back in for
merging with the upstream repo.

aps-7bm committed:
Major changes to 7-BM MEDM screens.
Add in ability to have a different bright exposure time.
Add in automatic shutter closing at the end of the scan for 7-BM.

aps-7bm committed:
Add screenshot of tomoScan_7BM.adl screen.

aps-7bm commented:
I added the screenshot of the MEDM screen, as requested.  I've kept working on this and have added a couple more features.  I added a different exposure time to the base class and screen, since that seemed like something that would be more generally useful.  I reworked my MEDM files quite a bit, since the 7BM screen was getting monstrous.  I also changed the logic a bit so that end_scan would get called in a finally: statement, so the end of scan tasks get done even if there is an exception thrown during the scan.

xray-imaging commented:
@aps-7bm all looks good to me. @MarkRivers may you please double check tomoscan/tomoscan.py changes?

MarkRivers commented on tomoScanApp/Db/tomoScan.template:
```diff
@@ -199,6 +199,17 @@ record(ao, "$(P)$(R)ExposureTime")
    field(PREC,  "3")
 }
 
+record(ao, "$(P)$(R)BrightExposureTime")
```
Everywhere else the term "FlatField" is used, not Briight.  This should be changed to FlatFieldExposureTime or if that is too long "FlatExposureTime".  Similarly it should be "DifferentFlatExposureTime".

MarkRivers replied:
Actually I think it should be "DifferentFlatExposure" without Time.  The PV name is getting near the maximum length depending on the prefixes.

MarkRivers commented on tomoScanApp/Db/tomoScan_settings.req:
```diff
@@ -68,6 +68,8 @@ $(P)$(R)SampleOutY
 # Exposure time
 #################
 $(P)$(R)ExposureTime
+$(P)$(R)BrightExposureTime
```
Change Bright to Flat here.

MarkRivers commented on tomoscan/tomoscan.py:
```diff
@@ -489,7 +493,27 @@ def set_exposure_time(self, exposure_time=None):
         if exposure_time is None:
             exposure_time = self.epics_pvs['ExposureTime'].value
-        self.epics_pvs['CamAcquireTime'].put(exposure_time)
+        self.epics_pvs['CamAcquireTime'].put(exposure_time, wait=True, timeout = 10.0)
```
Chnage Bright to Flat

MarkRivers commented on tomoScanApp/op/adl/tomoScan.adl:
```diff
@@ -1,14 +1,14 @@
 
 file {
```
I downloaded and viewed the new version of tomoScan.adl in medm.  It has a number of problems:
- The total height of the display is wrong, the bottom is cut off.  It was previously 825 but was changed to 803 for some reason.
- All widget X and Y positions should be a multiple of 5.  That keeps them correctly aligned. 
- Text labels should be positioned immediately adjacent to the widget they describe using "Space Evenly/Horizontal"
- Text labels should be exactly the size required by the text using "Size/Text to Contents".

MarkRivers commented:
I have made my suggested edits with medm to the adl file.  It is attached here with .txt on the end.
[tomoScan.adl.txt](https://github.com/tomography/tomoscan/files/5268222/tomoScan.adl.txt)

aps-7bm committed:
Suggested changes by Mark Rivers.  Change "Bright" to "Flat"
Revised tomoScan.adl screen to look better.

aps-7bm commented:
I made those changes and committed them, those were good points.  I should
have looked more carefully at your naming.  The MEDM screen looks better; I
frankly stink at making screens that look good, so I appreciate the
advice.  I have one question.  In the flat field part of the screen, all of
the labels on the far left are spaced a few pixels away from the fields to
their right except the Flat Exposure one.  Should we align it with the
labels above it, either by moving it left or the other labels right?

On Wed, Sep 23, 2020 at 7:58 AM Mark Rivers <notifications@github.com>
wrote:

> I have made my suggested edits with medm to the adl file. It is attached
> here with .txt on the end.
> tomoScan.adl.txt
> <https://github.com/tomography/tomoscan/files/5268222/tomoScan.adl.txt>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tomography/tomoscan/pull/81#issuecomment-697347311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOHRUIDXAQGXSEK5JGNB4DSHHWHFANCNFSM4RPN7RIQ>
> .
>

MarkRivers commented:
>  In the flat field part of the screen, all of the labels on the far left are spaced a few pixels away from the fields to their right except the Flat Exposure one. 

I don't see that, I just downloaded your latest version.  "Flat exposure" has X=35, and is 13 characters.  Each character takes 10 pixels in that font, so it ends at pixel 165.  The Menu widget to its right starts at pixel 170, so there are 5 pixels of space between them, which is what "Space evenly" does. 

Perhaps you are using a different font alias file from me?  This is what the screen looks like for me:

![image](https://user-images.githubusercontent.com/6115534/94022327-97a17600-fd7a-11ea-96d9-93e8f180806a.png)

There was one minor problem, those 2 widgets were 1 pixel too low.  That can be fixed with this patch:
```
corvette:tomoScanApp/op/adl>git diff .
diff --git a/tomoScanApp/op/adl/tomoScan.adl b/tomoScanApp/op/adl/tomoScan.adl
index 164a3c6..31f5eaf 100644
--- a/tomoScanApp/op/adl/tomoScan.adl
+++ b/tomoScanApp/op/adl/tomoScan.adl
@@ -1238,7 +1238,7 @@ menu {
 text {
        object {
                x=35
-               y=336
+               y=335
                width=130
                height=20
        }
@@ -1251,7 +1251,7 @@ text {
 menu {
        object {
                x=170
-               y=336
+               y=335
                width=100
                height=20
        }
```

aps-7bm commented:
Mark,
    This is what it looks like for me on my beamline workstation.

[image: image.png]

On Wed, Sep 23, 2020 at 8:56 AM Mark Rivers <notifications@github.com>
wrote:

> In the flat field part of the screen, all of the labels on the far left
> are spaced a few pixels away from the fields to their right except the Flat
> Exposure one.
>
> I don't see that, I just downloaded your latest version. "Flat exposure"
> has X=35, and is 13 characters. Each character takes 10 pixels in that
> font, so it ends at pixel 165. The Menu widget to its right starts at pixel
> 170, so there are 5 pixels of space between them, which is what "Space
> evenly" does.
>
> Perhaps you are using a different font alias file from me? This is what
> the screen looks like for me:
>
> [image: image]
> <https://user-images.githubusercontent.com/6115534/94022327-97a17600-fd7a-11ea-96d9-93e8f180806a.png>
>
> There was one minor problem, those 2 widgets were 1 pixel too low. That
> can be fixed with this patch:
>
> corvette:tomoScanApp/op/adl>git diff .
> diff --git a/tomoScanApp/op/adl/tomoScan.adl b/tomoScanApp/op/adl/tomoScan.adl
> index 164a3c6..31f5eaf 100644
> --- a/tomoScanApp/op/adl/tomoScan.adl
> +++ b/tomoScanApp/op/adl/tomoScan.adl
> @@ -1238,7 +1238,7 @@ menu {
>  text {
>         object {
>                 x=35
> -               y=336
> +               y=335
>                 width=130
>                 height=20
>         }
> @@ -1251,7 +1251,7 @@ text {
>  menu {
>         object {
>                 x=170
> -               y=336
> +               y=335
>                 width=100
>                 height=20
>         }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tomography/tomoscan/pull/81#issuecomment-697394881>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOHRUPMV4DJ6OPWY7SXVJDSHH5B7ANCNFSM4RPN7RIQ>
> .
>

MarkRivers commented:
@aps-7bm your screen-shot did not get uploaded properly.  You can drag a png file into the window, or copy and paste.

aps-7bm commented:
OK, that's weird.  When I look at it now, it looks just like yours.  Maybe
something didn't update properly earlier.  I'm good now, I think.  Any
other suggestions?

On Wed, Sep 23, 2020 at 9:36 AM Alan Kastengren <aps.7.bm@gmail.com> wrote:

> Mark,
>     This is what it looks like for me on my beamline workstation.
>
> [image: image.png]
>
> On Wed, Sep 23, 2020 at 8:56 AM Mark Rivers <notifications@github.com>
> wrote:
>
>> In the flat field part of the screen, all of the labels on the far left
>> are spaced a few pixels away from the fields to their right except the Flat
>> Exposure one.
>>
>> I don't see that, I just downloaded your latest version. "Flat exposure"
>> has X=35, and is 13 characters. Each character takes 10 pixels in that
>> font, so it ends at pixel 165. The Menu widget to its right starts at pixel
>> 170, so there are 5 pixels of space between them, which is what "Space
>> evenly" does.
>>
>> Perhaps you are using a different font alias file from me? This is what
>> the screen looks like for me:
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/6115534/94022327-97a17600-fd7a-11ea-96d9-93e8f180806a.png>
>>
>> There was one minor problem, those 2 widgets were 1 pixel too low. That
>> can be fixed with this patch:
>>
>> corvette:tomoScanApp/op/adl>git diff .
>> diff --git a/tomoScanApp/op/adl/tomoScan.adl b/tomoScanApp/op/adl/tomoScan.adl
>> index 164a3c6..31f5eaf 100644
>> --- a/tomoScanApp/op/adl/tomoScan.adl
>> +++ b/tomoScanApp/op/adl/tomoScan.adl
>> @@ -1238,7 +1238,7 @@ menu {
>>  text {
>>         object {
>>                 x=35
>> -               y=336
>> +               y=335
>>                 width=130
>>                 height=20
>>         }
>> @@ -1251,7 +1251,7 @@ text {
>>  menu {
>>         object {
>>                 x=170
>> -               y=336
>> +               y=335
>>                 width=100
>>                 height=20
>>         }
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/tomography/tomoscan/pull/81#issuecomment-697394881>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEOHRUPMV4DJ6OPWY7SXVJDSHH5B7ANCNFSM4RPN7RIQ>
>> .
>>
>

aps-7bm committed:
Minor change to align components in tomoScan.adl

MarkRivers commented:
It's OK to merge now in my opinion.

aps-7bm commented:
I just pushed that shift of the two components in the MEDM screen.

On Wed, Sep 23, 2020 at 10:38 AM Mark Rivers <notifications@github.com>
wrote:

> It's OK to merge now in my opinion.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tomography/tomoscan/pull/81#issuecomment-697558848>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOHRULQKUXFFDKTCARQWZLSHII5ZANCNFSM4RPN7RIQ>
> .
>

decarlof commented:
ok
